Project structure

danijelbuhin opened the pull request:
Currently covered:
 - Components organization (I kept focus only on the atomic structure since as far as I could tell, this is something we are leaning towards to)
 - Tests
 - Store
 - Fetchers


Inspiration:
- NLBL, soundpark, TPE, react-example and pd-vault repos
- [Atomic Design With React And Bit: Simplify a Complex UI](https://dev.to/giteden/atomic-design-with-react-and-bit-simplify-a-complex-ui-1k1f)
- [Atomic design methodology](https://atomicdesign.bradfrost.com/)

I did not wanted to go into very deep explanations to avoid a lot of text. Also, I did not cover some most common stuff like:
- Assets
- Locales
- Typings (interfaces, enums..)

If you think these points should be included, I could make section like `Misc`.

fvoska commented on React/NextJs/ProjectStructure.md:
```diff
@@ -0,0 +1,212 @@
+
+## Fetchers
+
+In `src/fetchers` you will organize your (swr) fetchers, usually by the model on which you will make API calls.
```
Not a big deal, but a bit strange folder hierarchy with models and fetchers.

Models are in `src/store/models` and fetchers that are in  `src/fetchers` are using those models. I would expect that the models and fetchers dirs are siblings (on the same level) instead of models being in a subdir.

Would one of the following options make sense?

1.
```
src/
  - store/
  - models/
  - fetchers/
```
2.
```
src/
  - store/
    - models/
    - fetchers/
```

DarkoKukovec replied:
I wouldn't mind if the models were in src - they are quite important and nesting them inside of the store folder is basically just mirroring a datx implementation detail, which is not necessary.

adrianbicanic replied:
Fetchers might not be related to the store (for example, if you are using `swr` to fetch google places), so I like the first, simplest option better. Also, it always bothered me that models are inside the store folder, although they are used globally (their purpose is not only for using them with the store, but also to get autocomplete inside components, utils, etc.).

fvoska commented on React/NextJs/ProjectStructure.md:
```diff
@@ -0,0 +1,212 @@
+src
+.
+└── components
+    └── ui
```
Are ALL atoms, molecules and organisms under top-level `components/ui` dir, or do you allow for multiple `ui` dirs for multiple features? For example, some page might have some organism component that is used only on that page. It wouldn't make sense to place that organism under global `components/ui`, I would place it in some dir like `page/my-page/components/ui/organism`. Basically allow to have this structure that you explained here both globally and for feature-specific cases.

The reason why I am saying this is that if an app is really large, you will have 100+ organism in `components/ui/organism` and some of them might be used only on certain pages/features and some are real globals.

DarkoKukovec replied:
@isBatak should probably give his feedback on this, but I believe we talked about the possibility of nesting components in such cases?
It would be good to explicitly mention it here.

adrianbicanic replied:
@DarkoKukovec Was there any idea of how to specifically nest such components, since they are not related to a context? 
@fvoska's example (`page/my-page/components/ui/organism`) also put's it into a context, which kind of beats the purpose ("At this level, the components begin to have the final shape, but they are still ensured to be independent, portable and reusable enough to be reusable in any content."), but in reality projects always get clogged and hard to read with a lot of files in a flat structure.

filipjakov replied:
@fvoska as quoted [here](https://github.com/infinum/JS-React-Example/blob/chakra/docs/atomic-design.md#organisms):

> Organisms are the combination of molecules that work together or even with atoms that compose more elaborate interfaces. At this level, the components begin to have the final shape, **but they are still ensured to be independent portable, and reusable enough to be reusable in any content**.

With that being said, organisms should still be reusable across different contexts/pages, and I believe their place is in the global ui folder.

I believe the mindset of context-bound (per page) organisms comes from one (or more) of the following situations:
- small project === small number of pages/features which results in a small number or organisms (no problem in traversing them at this point)
- component architecture is not setup as per the methodology violating the beforementioned principle (bolded in the quote)
- the design system is either bad and/or "versatile" resulting in the "inherent" creation of independent components. This might be out of scope here, but as projects get bigger there should be a consensus with the designers yadayada

Worst case scenario: you have a huge amount of code and you would like to scope the organisms in some manner; I'd propose to create subfolders which are scoped to either features or pages like so:

```
ui/
  organisms/
    feature1/
      some-organism
    feature2/
      some-organism
```

isBatak replied:
@fvoska  Per page component namespacing is kinda a wrong mindset here.
Organisms are basically "container" components but in the new React ecosystem, you don't have to fetch data at the top-level anymore, instead, you can fetch data in any component and still be able to reuse that component wherever you want. We collocate data fetching with a component to make it highly reusable, therefore you don't have to model your component based on the place where it is rendered, instead, you can model it by the data you are presenting.

This is the definition of the Organism that I found on the internet  
> Organisms are the combination of molecules that work together or even with atoms that compose more elaborate interfaces. At this level, the components begin to have the final shape, but they are still ensured to be independent, portable, and reusable enough to be reusable in any content.

I copied the whole article here, maybe we can reuse some definitions for our handbook.
https://github.com/infinum/JS-React-Example/blob/chakra/docs/atomic-design.md#organisms

[Templates](https://github.com/infinum/JS-React-Example/blob/chakra/docs/atomic-design.md#templates) and a good design system (like Chakra UI), kinda resolve this problem and need to scope components to a specific page. They shape the organisms to fit on a specific page... 

In conclusion, all of this is really new to us and we need to take it slow. 
I wouldn't go with per page namespacing because it seems like a wrong abstraction.

isBatak replied:
> React had a vision that goes beyond the browser. they understood that a clean component model is the unifying factor, that splitting up the view layer is a logical and technical mistake. the view and the code that forms it belong together, only state should be separate.

fvoska commented on React/NextJs/ProjectStructure.md:
```diff
@@ -0,0 +1,212 @@
+
+## Tests
+
+(For PR Review):
```
I would also add info about folder structure for tests for other units as well, like store, fetchers, etc.

We should also write some "recipes" about actually writing these tests for various types of units (components, hooks, store, fetchers, etc.). But those details are some other markdown file, perhaps `React/NextJs/Testing.md, this file is just for structure.

DarkoKukovec replied:
I agree, but this is maybe a topic for a different section/PR. It should be somewhere in between this one and https://github.com/infinum/frontend-handbook/pull/60

fvoska reviewed (changes requested):
Changed from Approved to Request Changes

DarkoKukovec commented on React/NextJs/ProjectStructure.md:
```diff
@@ -0,0 +1,212 @@
+
+### UI Components
+
+Before implementing [atomic design methodology](https://github.com/danilowoz/react-atomic-design), our common practice was separating components by the scope of the page + core components + shared components like this:
```
I'm not sure this has any value in the handbook since it will be obsolete and forgotten. Maybe it would be more useful as a comparison once the team starts using it.

DarkoKukovec commented on React/NextJs/ProjectStructure.md:
````diff
@@ -0,0 +1,212 @@
+      │   └── generateFileTree.test.ts
+      └── generateFileTree.ts
+
+```
````
Maybe a recap would be useful here, as a more complex example that includes most of the covered cases?

danijelbuhin commented:
Updates (synced with @mamihelj09): 
- Fixed component folder names
- Updated testing and styling sections

@isBatak and I will have a short meeting tomorrow about the extending atomic structure and overall progress of this handbook section, so more updates will soon come.

isBatak commented on React/NextJs/ProjectStructure.md:
```diff
@@ -0,0 +1,224 @@
+└── molecules
+    └── card
+        ├── layouts
+        │   ├── Desktop.tsx
```
Can we name it like this
`Card.desktop.tsx`, `Card.mobile.tsx`, `Card.tablet.tsx`
it is easier to search for a component.
And react by itself will introduce naming like this with server-side components, e.g. `Card.server.tsx` and `Card.client.tsx`

isBatak commented on React/NextJs/ProjectStructure.md:
```diff
@@ -0,0 +1,224 @@
+        ├── atoms
+        │   └── button
+        │       ├── Button.tsx
+        │       └── stories.tsx
```
We should name this `Button.stories.tsx` because it is easier to search through the project

danijelbuhin commented:
Latest update:

- Removed atomic design structure for components
- Changed the order of some sections to keep the reading flow meaningful

DarkoKukovec commented on React/NextJs/ProjectStructure.md:
```diff
@@ -30,64 +30,60 @@ src
+│       └── AdminBanner.tsx
+└── Gallery
+    ├── Gallery.tsx
+    └── components
```
Hmm, how would the nesting work here? Would we allow nesting of such components or would it only be on an "elements file" basis?

DarkoKukovec replied:
I'm not necessary against it, but it should be explicitly defined
